### ISSUE TITLE
External validation

### ISSUE THREAD

**Issue by brettz9** [gajus/eslint-plugin-jsdoc#493]
I think it could be useful for jsdoc to be usable for adding constraints externally to JavaScript source files (akin somewhat to a TypeScript declaration file or a JSON schema schema).

Note that this is not merely to check that jsdoc blocks are required, or that they are valid, but that they meet some pattern, e.g., a pattern to be enforced across multiple files. 

One use case includes using `overrides` in eslint config to ensure that all files in say a `plugins` folder export a default and/or named exports with a given signature.

For example, the config for plugins which act on, and return, a string might be:

```js
// .eslintrc.js
'use strict';
module.exports = {
  overrides: [
    {
      files: 'plugins/**',
      rules: {
        'jsdoc/check-signature': ['error', {
          defaultExport: `
            /**
              * @param {string} toConvert
              * @returns {string}
              */
          `
        }]
      }
    }
  ]
}
```

This config would then report if any JS files within `/plugins` failed to export a function with a single `toConvert` argument or return a string.

Named exports might be defined like:

```js
'jsdoc/check-signature': ['error', {
  namedExports: {
    user: '/** @type {string} */',
    port: '/** @type {number} */',
  }
}]
```

Along the lines of the above use case, such a rule might also be added as part of eslint configs/plugins which enforce various best practices and which also insist on particular exports.

A different use case, would be to avoid jsdoc taking up space inline, and might even be used to target validation of private methods and objects (in place of the `defaultExport` or `namedExports options, one might have a `namepaths` option, e.g., `{namepaths: {'SomeClass~privateMember': '/** @type {number}  */'}}`). Or maybe a `selectors` option based on https://eslint.org/docs/developer-guide/selectors#what-syntax-can-selectors-have.

(Other options might be used, e.g., to say that the signature need only apply *if and only if* such an export or namepath exists (and not otherwise), or to enforce partial signatures, e.g., check the first two arguments only or check that a type is used/returned but not that other types couldn't also be used.)

Although this type of validation might work with greater robustness with TypeScript (or with more preciseness in the case of JSON Schema when applied to JSON-friendly objects only), I think jsdoc has its place for such validation. It might even help transition to TypeScript. It could also supplement TypeScript, as, again, its intent is not exactly like a declaration file; it is not merely to define one's types as they are, but to indicate expectations for any new files.

**Update**: On further reflection, rather than `namedExports` and such, an object mapping from esquery selectors to the jsdoc would probably be more robust.

**Update 2**: https://github.com/dsheiko/bycontract/ looks helpful for consideration.

**Comment by brettz9** [comment]
While not of specific use to this plugin, I'll mention another use case for such jsdoc-based external validation that may be of interest in considering the utility of such external schemas: auto-complete.

When using an eslint rc file with the JavaScript extension (in order to allow code comments, shorter syntax, potentially dynamic settings), my IDE (Atom) can't get the benefit of autocomplete in showing allowable properties and values as it can using https://atom.io/packages/autocomplete-json (and its dependencies giving awareness of the eslint schema) with a JSON RC file (i.e., when using `.eslintrc.js` instead of `.eslintrc.json`).

If this were implemented, such a jsdoc-based schema (through `typedef`), could be usable to allow me to get auto-complete of `module.exports` properties within `.eslintrc.js` files. While TypeScript definitions could handle this as well, I think a jsdoc-based approach may be a lower barrier to entry.

**Comment by brettz9** [comment]
Closing since AST with the likes of `jsdoc/no-missing-syntax` and `jsdoc/no-restricted-syntax` can more precisely target specific structures.
